File: kakao.app/js/ked/view/menu/finder.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, lpad: function (l,s='',c=' ') {s=String(s); while(s.length<l){s=c+s} return s}, isNum: function (o) {return !isNaN(o) && !isNaN(parseFloat(o)) && (isFinite(o) || o === Infinity || o === -Infinity)}, min: function () { var m = Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.min.apply(_k_.min,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n < m ? n : m}}}; return m }, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}}

var finder

import kxk from "../../../kxk.js"
let post = kxk.post
let kseg = kxk.kseg
let kutil = kxk.kutil

import theme from "../../theme/theme.js"

import belt from "../../edit/tool/belt.js"

import inputchoice from "./inputchoice.js"


finder = (function ()
{
    _k_.extend(finder, inputchoice)
    function finder (screen, state, name = 'finder')
    {
        this.screen = screen
        this.state = state
    
        this["show"] = this["show"].bind(this)
        this["arrange"] = this["arrange"].bind(this)
        this["lineno"] = this["lineno"].bind(this)
        finder.__super__.constructor.call(this,this.screen,name,['gutter','scroll'])
        this.autoHideInput = false
        if (this.name === 'finder')
        {
            post.on('finder.show',this.show)
        }
        this.setColor('bg',theme.finder.bg)
        this.setColor('frame',theme.finder.frame)
        this.choices.state.skipAdjustViewForMainCursor = true
        this.choices.state.syntax.setExt('kode')
        this.choices.gutter.lineno = this.lineno
        this.choices.gutter.setColor('highlight',theme.gutter.fg)
        this.choices.gutter.setColor('bg_fully_selected',theme.finder.bg)
    }

    finder.prototype["lineno"] = function (y)
    {
        var pad, _47_43_

        pad = this.choices.gutter.cells.cols - 1
        if ((0 <= y && y < this.choices.fuzzied.length))
        {
            if (this.choices.fuzzied[y].type === 'file')
            {
                if ((this.choices.fuzzied[y].line != null))
                {
                    return _k_.lpad(pad,this.choices.fuzzied[y].line) + ' '
                }
                else
                {
                    return _k_.lpad(pad,'●') + ' '
                }
            }
            if (_k_.isNum(this.choices.fuzzied[y].row))
            {
                return _k_.lpad(pad,this.choices.fuzzied[y].row + 1) + ' '
            }
        }
        return _k_.lpad(pad + 1)
    }

    finder.prototype["arrange"] = function ()
    {
        var cs, h, w, x, y

        x = parseInt(this.screen.cols / 8)
        y = parseInt(this.screen.rows / 8)
        w = parseInt(this.screen.cols * 3 / 4)
        h = parseInt(this.screen.rows * 3 / 4 - 4)
        cs = _k_.min(h,this.choices.numFiltered())
        this.input.layout(x + 2,y + 1,w - 4,1)
        this.choices.layout(x + 1,y + 3,w - 3,cs)
        return this.cells.layout(x,y,w,cs + 4)
    }

    finder.prototype["show"] = function (text)
    {
        var cursorLine, front, span, _113_87_

        if (_k_.empty(text))
        {
            cursorLine = this.state.mainCursor()[1]
        }
        text = this.searchText(text)
        if (_k_.empty(text))
        {
            this.choices.clear()
            this.input.show()
            return finder.__super__.show.call(this)
        }
        this.state.highlightText(text)
        this.choices.clearEmpty()
        front = belt.frontmostSpans(this.state.s.highlights)
        var list = _k_.list(front)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            span = list[_a_]
            if (this.choices.items.length > 1 && this.choices.items.slice(-1)[0].row !== span[1] - 1)
            {
                this.choices.add({line:''})
            }
            this.choices.add({ext:'kode',line:' ' + kseg.str(this.state.s.lines[span[1]]),row:span[1],col:span[2]})
        }
        this.choices.state.highlightText(text)
        if (cursorLine)
        {
            this.choices.select(((_113_87_=kutil.findIndex(this.choices.items,function (l)
            {
                return l.row === cursorLine
            })) != null ? _113_87_ : 0))
        }
        else
        {
            this.choices.selectFirst()
        }
        this.input.show()
        this.input.grabFocus()
        return finder.__super__.show.call(this)
    }

    finder.prototype["searchText"] = function (text)
    {
        if (_k_.empty(text))
        {
            text = this.state.textOfSelectionOrWordAtCursor()
        }
        text = (text != null ? text : '')
        if (text !== this.input.current())
        {
            this.input.set(text)
            this.input.state.moveCursors('eol')
        }
        return text
    }

    finder.prototype["apply"] = function (choice)
    {
        if (!_k_.empty(choice))
        {
            post.emit('goto.line',choice.row,choice.col)
        }
        post.emit('focus','editor')
        this.hide()
        return {redraw:true}
    }

    finder.prototype["applyChoice"] = function (choice)
    {
        return this.apply(choice)
    }

    finder.prototype["onChoicesAction"] = function (action, choice)
    {
        switch (action)
        {
            case 'left':
                this.input.grabFocus()
                return true

        }

        return finder.__super__.onChoicesAction.call(this,action,choice)
    }

    finder.prototype["onInputAction"] = function (action, text)
    {
        switch (action)
        {
            case 'submit':
                return this.apply(this.choices.current())

            case 'change':
                if (!_k_.empty(text))
                {
                    return this.show(text)
                }
                else
                {
                    return this.choices.clear()
                }
                break
        }

        return finder.__super__.onInputAction.call(this,action,text)
    }

    return finder
})()

export default finder;

File: kakao.app/love/lua/view/menu/finder.lua
--[[
    ████████  ███  ███   ███  ███████    ████████  ████████   
    ███       ███  ████  ███  ███   ███  ███       ███   ███  
    ██████    ███  ███ █ ███  ███   ███  ███████   ███████    
    ███       ███  ███  ████  ███   ███  ███       ███   ███  
    ███       ███  ███   ███  ███████    ████████  ███   ███  

    searches and highlights in the current fileeditor file 
--]]

kxk = require "../../../kxk"
theme = require "../../theme"
tool = require "../../edit/tool"
 = require ""


local finder = class("finder", inputchoice)
    


function finder:init(@screen, @editor, name) 
        name = name or 'finder'
        
        self.state = self.editor.state discard procCall init(inputchoice(self), self.screen, name, array('gutter', 'scroll'))
        
        self.autoHideInput = false
        
        post.on('finder.show', if (self.name == 'finder') then self.show end)
        
        self:setColor('bg', theme.finder.bg)
        self:setColor('frame', theme.finder.frame)
        
        self.choices.state.skipAdjustViewForMainCursor = true
        self.choices.state.syntax.setExt('kode')
        self.choices.gutter.lineno = self.lineno
        -- revert gutter highlight color to default fg color since all lines are highlighted
        self.choices.gutter.setColor('highlight', theme.gutter.fg)
        self.choices.gutter.setColor('bg_fully_selected', theme.finder.bg)
        return self
    end

-- ███      ███  ███   ███  ████████  ███   ███   ███████ 
-- ███      ███  ████  ███  ███       ████  ███  ███   ███
-- ███      ███  ███ █ ███  ███████   ███ █ ███  ███   ███
-- ███      ███  ███  ████  ███       ███  ████  ███   ███
-- ███████  ███  ███   ███  ████████  ███   ███   ███████ 


function finder:lineno(y) 
        local pad = (self.choices.gutter.cells.cols - 1)
        if ((0 <= y) < #self.choices.fuzzied) then 
            if (self.choices.fuzzied[y].type == 'file') then 
                if self.choices.fuzzied[y].line then 
                    return (lpad(pad, self.choices.fuzzied[y].line) + ' ')
                else 
                    return (lpad(pad, '●') + ' ')
                end
            end
            
            if is(self.choices.fuzzied[y].row, num) then 
                return (lpad(pad, (self.choices.fuzzied[y].row + 1)) + ' ')
            end
        end
        
        return lpad((pad + 1))
    end

--  ███████   ████████   ████████    ███████   ███   ███   ███████   ████████
-- ███   ███  ███   ███  ███   ███  ███   ███  ████  ███  ███        ███     
-- █████████  ███████    ███████    █████████  ███ █ ███  ███  ████  ███████ 
-- ███   ███  ███   ███  ███   ███  ███   ███  ███  ████  ███   ███  ███     
-- ███   ███  ███   ███  ███   ███  ███   ███  ███   ███   ███████   ████████


function finder:arrangeRect() 
        return array(int(((self.editor.cells.x - self.editor.gutter.cells.cols) - 1)), int(self.editor.cells.y), int(((self.editor.cells.cols + self.editor.gutter.cells.cols) + 1)), int((self.editor.cells.rows - 3)))
    end


function finder:arrange() 
        array(x, y, w, h) = self:arrangeRect()
        
        local cs = min((h - 1), self.choices.numFiltered())
        
        self.input.layout((x + 2), (y + 1), (w - 4), 1)
        self.choices.layout((x + 1), (y + 3), (w - 2), cs)
        return self.cells.layout(x, y, w, (cs + 4))
    end

--  0000000  000   000   0000000   000   000  
-- 000       000   000  000   000  000 0 000  
-- 0000000   000000000  000   000  000000000  
--      000  000   000  000   000  000   000  
-- 0000000   000   000   0000000   00     00  


function finder:show(text) 
        -- log "finder.show #{@name} #{text}"
        
        if empty(text) then 
            local cursorLine = self.state.mainCursor()[1]
        end
        
        text = self:searchText(text)
        
        -- log g4("#{r6 'finder.show'} #{m5 @name} '#{text}'")
        
        if empty(text) then 
            self.choices.clear()
            self.input.show()
            return super()
        end
        
        self.state.highlightText(text)
        
        self.choices.clearEmpty()
        
        local front = belt.frontmostSpans(self.state.s.highlights)
        for span in front do 
            if ((#self.choices.items > 1) and (self.choices.items[-1].row ~= (span[1] - 1))) then 
                self.choices.add(line:'')
            end
            
            self.choices.add
            
                ext:'kode'
                (line:' ' + kseg.str(self.state.s.lines[span[1]]))
                row:span[1]
                col:span[2]
        end
        
        self.choices.state.highlightText(text)
        
        if cursorLine then 
            self.choices.select (kutil.findIndex(self.choices.items(l), function () (l.row == cursorLine) end) or 0)
        else 
            self.choices.selectFirst()
        end
        
        --inputFocus = @input.hasFocus()
        self.input.show()
        super() -- this might set focus to choices
        self.input.grabFocus() return -- but in finder and searcher input should always be focused
    end

--  0000000  00000000   0000000   00000000    0000000  000   000        000000000  00000000  000   000  000000000  
-- 000       000       000   000  000   000  000       000   000           000     000        000 000      000     
-- 0000000   0000000   000000000  0000000    000       000000000           000     0000000     00000       000     
--      000  000       000   000  000   000  000       000   000           000     000        000 000      000     
-- 0000000   00000000  000   000  000   000   0000000  000   000           000     00000000  000   000     000     


function finder:searchText(text) 
        -- log "finder.searchText #{@name} #{text}"
        
        if empty(text) then 
            -- log "finder.searchText #{@name} take text of selection or word at cursor"
            text = self.state.textOfSelectionOrWordAtCursor()
        end
        
        text = text or ''
        
        if (text ~= self.input.current()) then 
            self.input.set(text)
            self.input.selectAll()
            self.input.state.moveCursors('eol')
        end
        
        return text
    end

--  0000000   00000000   00000000   000      000   000  
-- 000   000  000   000  000   000  000       000 000   
-- 000000000  00000000   00000000   000        00000    
-- 000   000  000        000        000         000     
-- 000   000  000        000        0000000     000     


function finder:apply(choice) 
        if valid(choice) then 
            post.emit('goto.line', choice.row, choice.col)
        end
        
        post.emit('focus', 'editor')
        -- log "#{@name} applyChoice -> hide"
        self:hide()
        return redraw:true
    end


function finder:applyChoice(choice) 
    return self:apply(choice)
    end

--  ███████  ███   ███   ███████   ███   ███████  ████████   ███████         ███████    ███████  █████████  ███   ███████   ███   ███
-- ███       ███   ███  ███   ███  ███  ███       ███       ███             ███   ███  ███          ███     ███  ███   ███  ████  ███
-- ███       █████████  ███   ███  ███  ███       ███████   ███████         █████████  ███          ███     ███  ███   ███  ███ █ ███
-- ███       ███   ███  ███   ███  ███  ███       ███            ███        ███   ███  ███          ███     ███  ███   ███  ███  ████
--  ███████  ███   ███   ███████   ███   ███████  ████████  ███████         ███   ███   ███████     ███     ███   ███████   ███   ███


function finder:onChoicesAction(action, choice) 
        -- log "finder.onChoicesAction #{action}"
        
        if (action == 'delete') then 
                self.input.state.delete('back')
                self.input.grabFocus()
                return
        elseif (action == 'left') then 
                self.input.grabFocus()
                return true
        end return discard procCall init(inputchoice(self), action, choice)
    end

-- ███  ███   ███  ████████   ███   ███  █████████         ███████    ███████  █████████  ███   ███████   ███   ███
-- ███  ████  ███  ███   ███  ███   ███     ███           ███   ███  ███          ███     ███  ███   ███  ████  ███
-- ███  ███ █ ███  ████████   ███   ███     ███           █████████  ███          ███     ███  ███   ███  ███ █ ███
-- ███  ███  ████  ███        ███   ███     ███           ███   ███  ███          ███     ███  ███   ███  ███  ████
-- ███  ███   ███  ███         ███████      ███           ███   ███   ███████     ███     ███   ███████   ███   ███


function finder:onInputAction(action, text) 
        if (action == 'submit') then return self:apply(self.choices.current())
        elseif (action == 'change') then 
                if valid(text) then 
                    return self:show(text)
                else 
                    return self.choices.clear()
                end
        end return discard procCall init(inputchoice(self), action, text)
    end

export finder